text/0477-blueprints-update.md

```diff
@@ -0,0 +1,248 @@
+
+## Motivation
+
+Projects like [libkit](https://github.com/tildeio/libkit), [@glimmer/blueprint](https://github.com/glimmerjs/glimmer.js/tree/master/packages/%40glimmer/blueprint), [@ember/octane-app-blueprint](https://github.com/ember-cli/ember-octane-blueprint), and other app replacement blueprints become out-of-date instantly after generation. Tweaks to the blueprints are only beneficial to newly generated apps.
+
+## Detailed design
+
+**State storage schema**
+
+Up until now, ember-cli-update could update apps without any configuration, but in order to handle all blueprints and their options, some state has to be stored in the project.
+
+The code is already completed https://github.com/ember-cli/ember-cli-update/pull/552, but I wanted to publish this RFC before releasing it into the wild. I want to make sure the format of these files are agreed upon, because if this is successful, it will probably propagate throughout the ecosystem.
+
+Where should the state go?
+
+```js
+// .ember-cli
+{
+  "ember-cli-update": {
+    "blueprints": [
+      {
+        "name": "my-custom-blueprint",
+        "version": "0.0.1"
+      }
+    ]
+  }
+}
+```
+
+```js
+// ember-cli-update.json
+{
+  "blueprints": [
+    {
+      "name": "my-custom-blueprint",
+      "version": "0.0.1"
+    }
+  ]
+}
+```
+
+```js
+// package.json
+{
+  // ...
+  "ember-cli-update": {
+    "blueprints": [
+      {
+        "name": "my-custom-blueprint",
+        "version": "0.0.1"
+      }
+    ]
+  }
+}
+```
+
+I vote “ember-cli-update.json”. Since the code will be modified by the updater, it’s easier to reformat JSON. “.ember-cli” I think supports JS, so it would be hard to modify. “package.json” seems like an abuse to put more metadata in there.
+
+I've found via further testing that mixing a file that is tracked by the default ember-cli blueprint (.ember-cli, package.json) and modified by ember-cli-update (to update the blueprint metadata), it gets cumbersome. Both processes try to edit the same file. So in theory, ember-cli could alter the .ember-cli file in a way that conflicts with what you have in yours. This would give you a git conflict, then the updater can't read it anymore because it is in a state of invalid JS. I think this further supports a separate file.
+
+How should the state be structured?
+
+```js
+{
+  "blueprints": [
+    {
+      "name": "my-custom-blueprint",
+      "version": "0.0.1"
+    }
+  ]
+}
+```
+
+```js
+{
+  "blueprints": {
+    "my-custom-blueprint": {
+      "version": "0.0.1"
+    }
+  }
+}
+```
+
+Future-proofing for more options
+
+```js
+{
+  "additionalFutureOptions": null,
+  "blueprints": {
+    "my-custom-blueprint": {
+      "version": "0.0.1",
+      "additionalFutureOptions": null,
+    }
+  }
+}
+```
+
+**Methods of delivery**
+
+Blueprints could be responsible for injecting the state into projects
+
+```js
+// my-custom-blueprint/blueprints/my-custom-blueprint/files/.ember-cli
```

Comment:
Let's consolidate all examples in this RFC on the `ember-cli-update.json` format. I find it pretty confusing to jump amongst 2 or 3 different formats while reviewing.